lib/wallaroo/core/common/state_step_creator.pony

```diff
@@ -0,0 +1,286 @@
+    _pending_steps.create()
+
+  var _router_registry: (None | RouterRegistry) = None
+
+  new create(auth: AmbientAuth,
+    app_name: String,
+    worker_name: String,
+    metrics_conn: MetricsSink,
+    event_log: EventLog)
+  =>
+    _auth = auth
+    _app_name = app_name
+    _worker_name = worker_name
+    _metrics_conn = metrics_conn
+    _event_log = event_log
+
+  be application_begin_reporting(initializer: LocalTopologyInitializer) =>
+    initializer.report_created(this)
+
+  be application_created(initializer: LocalTopologyInitializer,
+    omni_router: OmniRouter)
+  =>
+    _omni_router = omni_router
+
+  be application_initialized(initializer: LocalTopologyInitializer) =>
+    initializer.report_ready_to_work(this)
+
+  be application_ready_to_work(initializer: LocalTopologyInitializer) =>
+    None
+
+  be report_unknown_key(producer: Producer, state_name: String, key: Key) =>
+    if not _keys_to_steps.contains(state_name, key) then
+      @printf[I32]("State Step Creator should create step for key '%s'\n"
+        .cstring(), key.cstring())
+
+      try
+        (let recovery_replayer, let omni_router) = try
+          (_recovery_replayer as RecoveryReplayer,
+            _omni_router as OmniRouter)
+        else
+          @printf[I32]("Missing things in StateStepCreator\n".cstring())
+          Fail()
+          error
+        end
+
+        let reporter = MetricsReporter(_app_name, _worker_name, _metrics_conn)
+
+        let runner_builder = try
+          _runner_builders(state_name)?
+        else
+          @printf[I32]("Could not find runner_builder for state '%s'\n".cstring(),
+            state_name.cstring())
+          Fail()
+          error
+        end
+
+        let id = _step_id_gen()
+        let state_step = Step(_auth, runner_builder(
+          where event_log = _event_log, auth = _auth),
+          consume reporter, id, runner_builder.route_builder(),
+          _event_log, recovery_replayer, _outgoing_boundaries,
+          this
+          where omni_router = omni_router)
+
+        _pending_steps(state_step) = (state_name, key, id)
+        state_step.initializer_initialized(this)
+      end
+    end
+
+  be set_router_registry(router_registry: RouterRegistry) =>
+    _router_registry = router_registry
+
+  be initialize_routes_and_builders(initializer: LocalTopologyInitializer,
+    keys_to_steps: KeyToStepInfoTag[Step] iso,
+    keys_to_step_ids: KeyToStepInfo[StepId] iso,
+    recovery_replayer: RecoveryReplayer,
+    outgoing_boundaries: Map[String, OutgoingBoundary] val,
+    runner_builders: Map[String, RunnerBuilder] val)
+  =>
+    _keys_to_steps = consume keys_to_steps
+    _keys_to_step_ids = consume keys_to_step_ids
+
+    _recovery_replayer = recovery_replayer
+    _outgoing_boundaries = outgoing_boundaries
+
+    _runner_builders = runner_builders
+
+    initializer.report_initialized(this)
+
+  be report_ready_to_work(initializable: Initializable) =>
```

Comment:
Why `Initializable` instead of `Step` here?

Reply:
No reason. Changed.